### ISSUE TITLE
ESLint should control plugin use by extension, not plugins

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#5155]
Per discussion in #5136, I am opening a separate issue.

It is currently not possible to dynamically configure the extensions a plugin will operate on. Currently, a plugin exports rules and processors:

```js
module.exports = {
    rules: {
        // ...
    },
    processors: {
        ".txt": someProcessor,
        ".html": someOtherProcessor
    }
};
```

As you can see, the plugin is expected to export an object, so it has no way of accessing settings or any dynamic information at that level. This implies that the extensions the plugin will process must also be static/constant.

@nzakas points out below that ESLint (or rather, the author of the config file used by ESLint) should control what plugins are applied to what extensions. So the goal here is to support ESLint-defined extension configuration, while allowing plugin-defined extension configuration as a fallback until 3.0.

Regarding the issue that plugins could "fight with each other" for the same file extensions: This can already happen in ESLint even without this proposal. It is true that it would be good for plugins to be able to decide not to handle a file at runtime, rather than just statically based on the extension (which would allow plugins that only handle certain files to decide to "pass on" a file and let a different plugin handle it). But this is mitigated by at least allowing the user better control.

**Comment by eslintbot** [comment]
@platinumazure Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
I think the way we setup processors today is backwards. I take full responsibility because (I believe) it was my idea to have plugins export the extensions that processors apply to. That's an inversion of responsibility that just doesn't make sense. In no other way can plugins impose behavior on an ESLint run.

What I'd rather do, instead of doubling-down on this broken plugin processor model, is look towards switching to a more sane model like the one Karma uses. So you'd need to configure it in your `.eslintrc.*` file, such as:

```
rules:
   foo: 1
processors:
  ".md": "myplugin/markdown"
```

**Comment by platinumazure** [comment]
@nzakas Before I say anything else, I just wanted to thank you for being upfront about the previous design decision. It takes a lot of guts to do that.

Okay, I think I can see how this can be done with a backward compatibility layer until 3.0. It's not pretty, though...

1. CLIEngine should build an extension-to-process mapping earlier in the process (no need to iterate through all plugins every time and it shouldn't change during a run, except due to configuration cascading.
1. When building this map, we should first look for `config.processors`. If not found there, revert to old behavior.
1. Then `processFile` in CLIEngine should be able to get a processor quickly without needing to worry about configuration.
1. In 3.0.0, we can remove support for processor-provided extension maps.

Does this sound remotely feasible?

**Comment by nzakas** [comment]
Sounds like a good direction. We may want to consider using a full glob rather than just extension.

**Comment by platinumazure** [comment]
@nzakas If we want to think about full glob configuration, maybe this discussion should be folded into #3611? What do you think?

**Comment by nzakas** [comment]
Hmm, perhaps. I suppose the question is if we want one way to configure globs at the top level or multiple levels.

**Comment by platinumazure** [comment]
What do users want? I mean, personally I'm not a huge fan of having multiple configuration files, but I'm also not super familiar with globbing and until recently would have assumed that there was no good way to do configuration in one file. Now I'm beginning to think it could be doable... but that's just my experience/opinion. Probably worth asking the broader community somehow.

**Comment by JustinLivi** [comment]
@platinumazure This comment may be more relevant in #3611 but is an attempt to directly answer you question. As a user, I've recently found myself in the situation where a single configuration based on globbing is looking more and more enticing/necessary. I'm working on a express/react application with multiple routes and I'd really like to structure my application like this:

```
project-root
├── .eslintrc
└── src
    ├── components
    │   ├── foo.js
    │   └── bar.js
    └── routes
        ├── routeFoo
        │   ├── client.js
        │   └── server.js
        └── routeBar
            ├── client.js
            └── server.js
```

Ideally I'd like to have one configuration applied to all client.js files under src/routes as well as all the files under components. Then I'd like a separate configuration applied to all server.js files under src/routes.

A lot of the rules will apply to both of my configurations, so what I'd really like to see is the cascading functionality *within a single .eslintrc file*:
```
{
    "rules": { "indent": [ 2, 4, { "SwitchCase": 1 }] },
    "children": [
        {
            "pattern": ["src/components/**/*.js", "src/routes/**/client.js"],
            "ecmaFeatures": { "modules": true },
        },
        {
            "pattern":  "src/routes/**/server.js",
            "ecmaFeatures": { "modules": false },
        }
    ]
}
```

The proposal would add two additional properties to the config
1. an optional "pattern" glob property to which the entire parent rule tree would be applied
2. an optional "children" property which would be an array of children configurations. These would cascade in the same manner that actual .eslintrc files do currently

**Comment by platinumazure** [comment]
Closing due to lack of activity and because glob-based config will fix this, so this is sort of a duplicate.
